File: 프로그래머스/2/87946. 피로도/피로도.py
answer = 0
visited = []

def backtrack(k, dungeons, cnt):
    global answer
    answer = max(answer, cnt)
    
    for i in range(len(dungeons)):
        if not visited[i] and k>=dungeons[i][0]:
            visited[i] = True
            backtrack(k-dungeons[i][1], dungeons, cnt+1)
            visited[i] = False

def solution(k, dungeons):
    global visited
    visited = [False] * len(dungeons)
    backtrack(k, dungeons, 0)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer = -1;
    boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
    
        return answer;
    }
    
    private void dfs(int currK, int[][] dungeons, int count) {
        answer = Math.max(answer, count);
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && currK >= dungeons[i][0]) {
                visited[i] = true;
                dfs(currK - dungeons[i][1], dungeons, count+1);
                visited[i] = false;
            }
        }
    }
}